drivers/wifi/esp8266/wifi_esp8266.c

```diff
@@ -0,0 +1,1452 @@
+	if (ret < 0) {
+		ret = -EINVAL;
+		goto ret_early;
+	}
+
+	foo_data.sending_data = 0;
+	k_sem_reset(&foo_data.response_sem);
+	k_mutex_lock(&dev_mutex, K_FOREVER);
+	while (send_len && frag) {
+		if (frag->len > send_len) {
+			mdm_receiver_send(&foo_data.mdm_ctx, frag->data,
+					  send_len);
+
+			send_len = 0;
+			break;
+		} else {
+			mdm_receiver_send(&foo_data.mdm_ctx, frag->data,
+					  frag->len);
+			send_len -= frag->len;
+			frag = frag->frags;
+		}
+	}
+	k_mutex_unlock(&dev_mutex);
+
+	ret = k_sem_take(&foo_data.response_sem, MDM_CMD_TIMEOUT*2);
+
+	if (ret == 0) {
+		ret = foo_data.last_error;
+	} else if (ret == -EAGAIN) {
+		ret = -ETIMEDOUT;
+	}
+
+ret_early:
+	net_pkt_unref(pkt);
+	if (cb) {
+		cb(context, ret, user_data);
+	}
+
+	return ret;
+}
+
+static int esp8266_put(struct net_context *context)
+{
+	struct socket_data *data;
+	u8_t msg[30];
+	int len;
+
+	if (!context || !context->offload_context) {
+		return -EINVAL;
+	}
+
+	data = context->offload_context;
+	int id = data - foo_data.socket_data;
+	foo_data.sock_map &= ~(1 << id);
+
+	foo_data.socket_data[id].recv_cb = NULL;
+	foo_data.socket_data[id].send_cb = NULL;
+
+	if (foo_data.socket_data[id].connected) {
+		len = sprintf(msg, "AT+CIPCLOSE=%d", id);
+		msg[len] = '\0';
+
+		if (send_at_cmd(&foo_data.socket_data[id], msg,
+				MDM_CMD_TIMEOUT)) {
+			LOG_ERR("failed to close\n");
+		}
+	}
+
+	net_context_unref(context);
+
+	data->context = NULL;
+	memset(&data->src, 0, sizeof(struct sockaddr));
+	memset(&data->dst, 0, sizeof(struct sockaddr));
+	return 0;
+}
+
+static int esp8266_send(struct net_pkt *pkt,
+		net_context_send_cb_t cb,
+		s32_t timeout,
+		void *user_data)
+{
+	struct net_context *context = net_pkt_context(pkt);
+	socklen_t addrlen;
+	int ret;
+
+	addrlen = 0;
+	if (net_pkt_family(pkt) == AF_INET) {
+		addrlen = sizeof(struct sockaddr_in);
+	} else {
+		return -EPFNOSUPPORT;
+	}
+
+	ret = esp8266_sendto(pkt, &context->remote, addrlen, cb,
+			      timeout, user_data);
+	return ret;
+}
+
+static int esp8266_recv(struct net_context *context,
+		net_context_recv_cb_t cb,
+		s32_t timeout,
+		void *user_data)
+{
+	struct socket_data *data;
+
+	if (!context || !context->offload_context) {
+		return -EINVAL;
+	}
+
+	data = context->offload_context;
+
+	data->recv_cb = cb;
+	data->recv_user_data = user_data;
+
+	return 0;
+}
+
+static struct net_offload esp8266_offload = {
+	.get            = esp8266_get,
+	.bind		= esp8266_bind,
+	.listen		= esp8266_listen,
+	.connect	= esp8266_connect,
+	.accept		= esp8266_accept,
+	.send		= esp8266_send,
+	.sendto		= esp8266_sendto,
+	.recv		= esp8266_recv,
+	.put		= esp8266_put,
+};
+
+static int esp8266_mgmt_scan(struct device *dev, scan_result_cb_t cb)
+{
+	int ret;
+
+	foo_data.wifi_scan_cb = cb;
+
+	ret = send_at_cmd(NULL, "AT+CWLAP", MDM_CMD_TIMEOUT);
+	if (ret < 0) {
+		LOG_ERR("failed to send scan\n");
+		ret = -EINVAL;
+	}
+
+	foo_data.wifi_scan_cb = NULL;
+	return 0;
+};
+
+static u8_t connect_msg[100];
```

Comment:
Perhaps this should be part of the `struct esp8266_data` that is defined in the beginning of the file.